### ISSUE TITLE
Helm causes "M-x gdb" to freeze emacs in Mac OS X when gdb finishes debugging

### ISSUE THREAD

**Issue by oracleyue** [emacs-helm/helm#1168]
For details of how this bug looks, refer to
http://stackoverflow.com/questions/28834723/gdb-emacs-hangs-after-running-any-binary/32556798#32556798

I guess it could be very difficult to track, since it works quite well in Linux (having tested in Arch Linux). The bug only appears in Mac OS X. Any ideas? It may not be a serious bug, you might track it when you got time.

Many thanks!

-------
Quick procedure to repeat the bug:
- write a simple C program "main.c", compile it by `gcc -g -o foo main.c`: 
```c
#include<stdio.h>
int main()
{
    int n =0;
    printf("Hello, world!");
}
```
- In emacs, cd to the folder of `foo`, `M-x gdb`; then in the gdb buffer, type in `b main` hit Enter, `r` hit Enter; it should run well.
- Then type `c`, hit Enter, you will see emacs hangs, with 100% cpu consumption.

----------

**Comment by thierryvolpiatto** [comment]
Zuogong Yue <notifications@github.com> writes:

> For details of how this bug looks, refer to
> http://stackoverflow.com/questions/28834723/gdb-emacs-hangs-after-running-any-binary/32556798#32556798
>
> I guess it could be very difficult to track, since it works well in
> Linux (having tested in Arch Linux). The bug only appears in Mac OS X.

What make you believe that it is helm the cause of this bug ?

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by oracleyue** [comment]
@thierryvolpiatto 
I cannot be sure about who should be "blamed" on this issue, `helm` or `gdb-mi`. 

FYI,

At the beginning, I have no idea what causes this bug, and I thought it is a bug of `gdb-mi` (probably there are also bugs in `gdb-mi`).

Then I disabled all configurations in my `.emacs`, `M-x gdb` works well. Later I enable modes almost one by one, I found that once I disabled `helm` and other modes depending on `helm` (`helm-swoop`, `helm-projectile`, `function-args`), `M-x gdb` will work without this issue. Once I enabled `helm` (keeping disabled the minor modes that depend on `helm`), `M-x gdb` produced the issue.

Thus, I guess there is a conflict between `helm` and `gdb-mi`. You are right, it is hard to say it is a bug of `helm`, instead of `gdb-mi`.

**Comment by thierryvolpiatto** [comment]
Zuogong Yue <notifications@github.com> writes:

> @thierryvolpiatto
> I cannot be sure about who should be "blamed" on this issue, helm or gdb-mi.
>
> FYI,
>
> At the beginning, I have no idea what causes this bug, and I thought
> it is a bug of gdb-mi (probably there are also bugs in gdb-mi).
>
> Then I disabled all configurations in my .emacs, M-x gdb works
> well. Later I enable modes almost one by one, I found that once I
> disabled helm and other modes depending on helm (helm-swoop,
> helm-projectile, function-args), M-x gdb will work without this
> issue. Once I enabled helm (keeping disabled the minor modes that
> depend on helm), M-x gdb produced the issue.
>
> Thus, I guess there is a conflict between helm and gdb-mi. You are
> right, it is hard to say it is a bug of helm, instead of gdb-mi.

The only place I see in gdb-mi where helm-mode could be involved is a
completing-read in `gud-watch', and even if I never used this, I hardly
see why it would be involved when gdb finish.
I am speaking only for helm-mode, I am not responsible and fan 
of other modes mentioned.

Perhaps try to run gdb-mi in ./emacs-helm.sh to see if you can reproduce. 

Unfortunately I have no access to MacOSx...

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by oracleyue** [comment]
@thierryvolpiatto 
>Perhaps try to run gdb-mi in ./emacs-helm.sh to see if you can reproduce.

Could you tell me a bit more how to perform your suggestion? 
Where is `emacs-helm.sh`? What do you mean run `gdb-mi` in `emacs-helm.sh`?

Many thanks!

**Comment by thierryvolpiatto** [comment]
Zuogong Yue <notifications@github.com> writes:

> @thierryvolpiatto
>
>     Perhaps try to run gdb-mi in ./emacs-helm.sh to see if you can reproduce.
>    
> Could you tell me a bit more how to perform your suggestion?
> Where is emacs-helm.sh? What do you mean run gdb-mi in emacs-helm.sh?

emacs-helm.sh is part of helm.

cd /your/helm/directory
./emacs-helm.sh -P /path/to/your/emacs

Then when emacs is started try to reproduce your bug.

NOTE: if emacs in in PATH, you can just run as we do on GNU/Linux:

./emacs-helm.sh

But it seems people on Macosx generally need to specify the path of
emacs with the "-P" option, (see ./emacs-helm.sh -h).

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by oracleyue** [comment]
@thierryvolpiatto 

I run `./emacs-helm.sh`, it runs well. Then I test `gdb-mi`, it works well, without freezing emacs. 

Is it means this is not a bug of `helm`? If `emacs-helm.sh` includes all features of `helm`, it is quite a strong evidence for me.

Okay, probably I need to test other modes or configurations that cause the issue.

Thanks a lot!

**Comment by thierryvolpiatto** [comment]
Zuogong Yue <notifications@github.com> writes:

> @thierryvolpiatto
>
> I run ./emacs-helm.sh, it runs well. Then I test gdb-mi, it works well, without freezing emacs.
>
> Is it means this is not a bug of helm?

Yes.

> Okay, probably I need to test other modes or configurations that cause the issue.

Just a guess, have a look at what does find-file-hook...

-- 
Thierry
https://emacs-helm.github.io/helm/

**Comment by oracleyue** [comment]
@thierryvolpiatto 

I finally located the bug. It is caused by CEDET again. Its semantic-mode has lots of bugs in Mac OS X, even the basic functionality: parsing system C/C++ library. In the configuration .el of helm, I want to use `helm-semantic-or-imenu`, so I enable `(semantic-mode 1)`. (`imenu` doesn't show results as fancy as `semantic`).

Once I disabled all packages in CEDET (I only enabled `semantic-mode`), everything goes well now.
